File: Practise/ArrayHashing/Patterns/TwoPointer/2Sum/2sum.py
from typing import List, Dict

def two_sum(arr : List[int], target : int) -> List[int]:
    num_map : Dict[int,int] = {}
    
    for i,num in enumerate(arr):
        # Ans found condition
        
        if num in num_map:
            return[num_map[num],i]
        
        num_map[target - num] = i
    
    return []
    # Ans not found    
arr : List[int] = [2,7,11,15]
target : int = 9
print(f"Indicises of two numbers : {two_sum(arr,target)}")

File: Practise/ArrayHashing/Patterns/TwoPointer/2Sum/2sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

vector<int> two_sum(const vector<int>& arr,int target)
{
    unordered_map<int, int> umap;
    for (int i = 0; i < arr.size() ; i++)
    {
        // Stopping condition / answer
        if(umap.find(arr[i]) != umap.end())
        {
            return {umap[arr[i]], i};
        }

        umap[target - arr[i]] = i;
    }

    // No answer
    return {};
}

int main()
{
    std::vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    std::vector<int> result = two_sum(nums, target);
    std::cout << "Indices of the two numbers: [" << result[0] << ", " << result[1] << "]\n";

    return 0;
}

